ext/atomic_ruby/atomic_ruby.c

```diff
@@ -62,14 +73,17 @@ static VALUE rb_cAtom_swap(VALUE self) {
   do {
     expected_old_value = atomic_ruby_atom->value;
     new_value = rb_yield(expected_old_value);
+#ifdef HAVE_RACTOR_SHAREABLE_PROC
+    check_value_shareable(new_value);
+#endif
   } while (RUBY_ATOMIC_VALUE_CAS(atomic_ruby_atom->value, expected_old_value, new_value) != expected_old_value);
   RB_OBJ_WRITTEN(self, expected_old_value, new_value);
 
   return new_value;
 }
 
 RUBY_FUNC_EXPORTED void Init_atomic_ruby(void) {
-#ifdef HAVE_RB_EXT_RACTOR_SAFE
+#if defined(HAVE_RB_EXT_RACTOR_SAFE) && defined(HAVE_RACTOR_SHAREABLE_PROC)
   rb_ext_ractor_safe(true);
 #endif
```

Comment:
I think it would make sense to keep this one as-is, and rather only do `| RUBY_TYPED_FROZEN_SHAREABLE` above if checking if the value is shareable.
BTW, why* not always check that the value if shareable, regardless of `Ractor.respond_to?(:shareable_proc)`?

Reply:
> BTW, why not always check that the value is shareable, regardless of Ractor.respond_to?(:shareable_proc) and version?

Because I originally released this for versions 3.3 and above, where there's no easy way to make a proc shareable afaik, but the thread pool's design relies on callable work being passed into it. So, I'd rather only do these Ractor safety checks on 3.5+, where `Ractor.shareable_proc` is available. I'll probably just do a version check though and simplify all of this.

Reply:
> To be safe, one needs to always check that the value if shareable, for assigning to a `RUBY_TYPED_FROZEN_SHAREABLE` object. See [ruby/ruby@`master`/doc/extension.rdoc#appendix-f-ractor-support-](https://github.com/ruby/ruby/blob/master/doc/extension.rdoc?rgh-link-date=2025-10-31T16%3A52%3A02Z#appendix-f-ractor-support-) specifically points 1 and 4:
> 
> > 1. Do not share unshareable objects between ractors
> > 2. RUBY_TYPED_FROZEN_SHAREABLE flag indicates that these objects can be shareable objects if the object is frozen. This means that if the object is frozen, the mutation of wrapped data is not allowed.
> 
> (mutation of wrapped data is what I'm doing in the bug repro script below)

Noted. I'll also be setting that flag only on 3.5+ 👍🏽